cc/mcts_node.cc

```diff
@@ -0,0 +1,254 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "cc/mcts_node.h"
+
+#include <algorithm>
+#include <cmath>
+#include <functional>
+#include <iomanip>
+#include <sstream>
+#include <tuple>
+#include <utility>
+
+#include "cc/algorithm.h"
+
+namespace minigo {
+
+constexpr float MctsNode::kPuct;
+
+MctsNode::MctsNode(EdgeStats* stats, const Position& position)
+    : parent(nullptr), stats(stats), move(Coord::kInvalid), position(position) {
+  for (int i = 0; i < kNumMoves; ++i) {
+    illegal_moves[i] = position.IsMoveLegal(i) ? 0 : 1000;
+  }
+  DualNet::InitializeFeatures(position, &features);
+}
+
+MctsNode::MctsNode(MctsNode* parent, Coord move)
+    : parent(parent),
+      stats(&parent->edges[move]),
```

Comment:
Yep, that's correct.

I ended up having to write it this way to handle the root node case, because there's no easy way to replicate DummyNode from the Python implementation in C++.